boa_engine/src/object/builtins/jsnative_object.rs

```diff
@@ -0,0 +1,269 @@
+    ///
+    /// #[derive(Debug, Trace, Finalize)]
+    /// enum Animal {
+    ///     Cat,
+    ///     Dog,
+    ///     Other,
+    /// }
+    ///
+    /// // Implement native class for `Animal` through the `Class` trait.
+    /// impl Class for Animal {
+    ///       // See `Class` documentation for implementation.
+    ///       // ...
+    /// #     // We set the binging name of this function to be `"Animal"`.
+    /// #     const NAME: &'static str = "Animal";
+    /// #
+    /// #     // We set the length to `1` since we accept 1 arguments in the constructor.
+    /// #     const LENGTH: usize = 1;
+    /// #
+    /// #     // This is what is called when we do `new Animal()` to construct the inner data of the class.
+    /// #     fn make_data(_new_target: &JsValue, args: &[JsValue], context: &mut Context) -> JsResult<Self> {
+    /// #         // This is equivalent to `String(arg)`.
+    /// #         let kind = args.get_or_undefined(0).to_string(context)?;
+    /// #
+    /// #         let animal = match kind.to_std_string_escaped().as_str() {
+    /// #             "cat" => Self::Cat,
+    /// #             "dog" => Self::Dog,
+    /// #             _ => Self::Other,
+    /// #         };
+    /// #
+    /// #         Ok(animal)
+    /// #     }
+    /// #
+    /// #     /// This is where the object is initialized.
+    /// #     fn init(class: &mut ClassBuilder) -> JsResult<()> {
+    /// #         class.method(
+    /// #             js_string!("speak"),
+    /// #             0,
+    /// #             NativeFunction::from_fn_ptr(|this, _args, _ctx| {
+    /// #                 if let Some(object) = this.as_object() {
+    /// #                     if let Some(animal) = object.downcast_ref::<Animal>() {
+    /// #                         match &*animal {
+    /// #                             Self::Cat => println!("meow"),
+    /// #                             Self::Dog => println!("woof"),
+    /// #                             Self::Other => println!(r"¯\_(ツ)_/¯"),
+    /// #                         }
+    /// #                     }
+    /// #                 }
+    /// #                 Ok(JsValue::undefined())
+    /// #             }),
+    /// #         );
+    /// #         Ok(())
+    /// #     }
+    /// }
+    ///
+    /// # let context = &mut Context::default();
+    ///
+    /// // Create a `JsNativeObject<Animal>`
+    /// let js_dog = JsNativeObject::new(Animal::Dog, context);
+    /// ```
+    pub fn new(native_object: T, context: &mut Context) -> JsResult<Self>
+    where
+        T: Class,
+    {
+        let class = context.get_global_class::<T>().ok_or_else(|| {
+            JsNativeError::typ().with_message(format!(
+                "could not find native class `{}` in the map of registered classes",
+                T::NAME
+            ))
+        })?;
+
+        Ok(Self::new_with_proto(
+            class.prototype(),
+            native_object,
+            context,
+        ))
+    }
```

Comment:
Now that #3488 is merged, this functionality should be replaced by returning a `JsNativeObject` on `Class` itself. That should also ensure that any new additions to `Class` are correctly called; this implementation skips calling the `Class::object_constructor` method.